src/core/net/dnssd_server.cpp

```diff
@@ -1287,6 +1386,574 @@ void Server::ResetUpstreamQueryTransaction(UpstreamQueryTransaction &aTxn, Error
+{
+    DnsName instanceName;
+
+    VerifyOrExit(mIsRunning);
+    VerifyOrExit(aService.mInfraIfIndex == Get<BorderRouter::InfraIf>().GetIfIndex());
+
+    ConstructFullInstanceName(aService.mServiceInstance, aService.mServiceType, instanceName);
+    HandleResult(kResolvingService, instanceName, &Response::AppendService, ProxyResult(aService));
+
+exit:
+    return;
+}
+
+void Server::DiscoveryProxy::HandleIp6AddressResolveResult(Dnssd::Event aEvent, const Dnssd::Host &aHost)
+{
+    bool    hasValidAddress = false;
+    DnsName fullHostName;
+
+    VerifyOrExit(mIsRunning);
+    VerifyOrExit(aEvent == Dnssd::kEventAdded);
+    VerifyOrExit(aHost.mInfraIfIndex == Get<BorderRouter::InfraIf>().GetIfIndex());
+
+    for (uint16_t index = 0; index < aHost.mNumAddresses; index++)
+    {
+        const Ip6::Address &address = AsCoreType(&aHost.mAddresses[index]);
+
+        if (IsProxyAddressValid(address))
+        {
+            hasValidAddress = true;
+            break;
+        }
+    }
+
+    VerifyOrExit(hasValidAddress);
+
+    ConstructFullName(aHost.mHostName, fullHostName);
+    HandleResult(kResolvingIp6Address, fullHostName, &Response::AppendHostAddresses, ProxyResult(aHost));
+
+exit:
+    return;
+}
+
+void Server::DiscoveryProxy::HandleResult(ProxyAction        aAction,
+                                          const DnsName     &aName,
+                                          ResponseAppender   aAppender,
+                                          const ProxyResult &aResult)
+{
+    // Common method that handles result from DNS-SD on infra-if. It
+    // iterates over all `ProxyQuery` entries and checks if any entry
+    // is waiting for the result of `aAction` for `aName`. Matching
+    // queries are updated using the `aAppender` method pointer,
+    // which appends the corresponding records to the response. We
+    // then determine the next action to be performed for the
+    // `ProxyQuery` or if it can be finalized.
+
+    static const ProxyAction kNextAction[] = {
+        /* (0) kNoAction            -> */ kNoAction,
+        /* (1) kBrowsing            -> */ kResolvingService,
+        /* (2) kResolvingService    -> */ kResolvingIp6Address,
+        /* (3) kResolvingIp6Address -> */ kNoAction,
+    };
+
+    static_assert(0 == kNoAction, "kNoAction value is incorrect");
+    static_assert(1 == kBrowsing, "kBrowsing value is incorrect");
+    static_assert(2 == kResolvingService, "kResolvingService value is incorrect");
+    static_assert(3 == kResolvingIp6Address, "kResolvingIp6Address value is incorrect");
+
+    ProxyQueryList nextActionQueries;
+    ProxyQueryInfo info;
+    ProxyAction    nextAction;
+
+    for (ProxyQuery &query : Get<Server>().mProxyQueries)
+    {
+        Response response(GetInstance());
+        bool     shouldFinalize;
+
+        info.ReadFrom(query);
+
+        if (!QueryMatches(query, info, aAction, aName))
+        {
+            continue;
+        }
+
+        CancelAction(query, info);
```

Comment:
Yes. From what I recall, there were discussions on this (within the Thread group and during the initial implementation of the DNS-SD server module in OpenThread). This behavior was chosen to prioritize simplicity, ensure a quick response, and meet basic functionality requirements. I think one argument was that a client that wants to perform a continuous browse, should support and use other mechanisms (e.g. DNS Push).

Personally, I'm not fully in favor of this model. I think a short wait time model may be more preferable. This would allow the DNS-SD resolver to collect at least some browse results before sending the query response. This would be a larger change to the implementation but we may consider it at a later time.